expression/expr_to_pb.go

```diff
@@ -453,6 +458,22 @@ func (pc pbConverter) builtinFuncToPBExpr(expr *ScalarFunction) *tipb.Expr {
 	}
 }
 
+// bypassDeferredFuncToPBExpr does not convert a deferred function to a PB expression.
+// It will be used to check whether or not the expression including a deferred function
+// can be converted to a PB expression eventually.
+func (pc pbConverter) bypassDeferredFuncToPBExpr(expr *ScalarFunction) *tipb.Expr {
+	tp := tipb.ExprType_Null //FIXME: hack to bypass
```

Comment:
In fact, if you set `tipb.Expr.Tp` with `tipb.ExprType_Null`, TiKV will regard it as a NULL constant value and ignore all its parameters. This behavior pushed to TiKV is not correct, it may produce a wrong result when you push such a expression to TiKV. For example: `? > 12` will be represented by `NULL > 12` after pushed to TiKV

Currently, whether a function can be pushed down to TiKV is controlled by `pc.client.IsRequestTypeSupported` and `expr.Function.PbCode()`, see `convertToPBExpr` for more detail

It's ok to let function `getparam` or function that contains `getparam` not being pushed to TiKV. To do so, you can remove all the changes made in `expression/expr_to_pb.go`

Reply:
Thank you for your insightful comments. Should I introduce a special PBExpr to handle these expressions correctly? I think that the special PBExpr needs to represent a place holder for "getparam" and needs to be replaced with a real PBExpr of the value of getparam just before pusing down to TiKV.  

However, If I remove the all the changes as you suggested, the expressions including “getparam” will be in (3) “remained”.  That is, “getparam” expression is evaluated whenever fetching each tuple from TiKV, and accordingly the query performance might not be good. 

`30 func ExpressionsToPB(sc *variable.StatementContext, exprs []Expression, client kv.Client) (pbExpr *tipb.Expr, pushed []Expression, remained []Expression) {`
`31  pc := pbConverter{client: client, sc: sc}`
`32  for _, expr := range exprs {`
`33      v := pc.exprToPB(expr)`
`34      if v == nil {`
`35          remained = append(remained, expr)`
`36          continue`
`37      }`

Reply:
I think we can do the push down work in another pr.

In this pr, we can add a global or session variable, like `enable_plan_cache_for_prepare`, with default value `true`. If the performance is harmed, we can set the flag to `false` to disable the plan cache for prepared statement and remove all cache stored in every session.

After the push down work finished, this flag is still useful: for data increment, our CBO framework may chose another more optimized physical plan for some cached prepared statement.

Reply:
I've resolved the issues by introducing a special field type, TypeParam, which is only used internally.

TypeParam is for an unknown type, but its real type will be determined by other known types eventually. For example, the expression '1 + ?' will be converted to int_plus(int_constant('1'), int_cast(getparam(0))). For another example, the predicate expression 'int_key = ?' will be converted to eq(col:int_key, int_cast(getparam(0))).

For supporting the new distsqls executors, the dagPB of a cached plan will be regenerated whenever evaluating the parameters in the new executors. However, we can further optimize the regeneration of the dagPB by only replacing the parts changed by the parameter evaluations.

I've also added the global variable ,"enable-plan-cache-for-prepare", to set the flag of using the plan cache.